File: linkis-computation-governance/linkis-engineconn/linkis-clustered-engineconn/linkis-once-engineconn/src/main/scala/org/apache/linkis/engineconn/once/executor/service/JobHistoryService.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.engineconn.once.executor.service

import org.apache.linkis.common.utils.{Logging, Utils}
import org.apache.linkis.engineconn.executor.conf.EngineConnExecutorConfiguration
import org.apache.linkis.governance.common.entity.job.JobRequest
import org.apache.linkis.governance.common.protocol.job.{JobReqUpdate, JobRespProtocol}
import org.apache.linkis.protocol.message.RequestProtocol
import org.apache.linkis.rpc.Sender

import java.util.Date

/**
 * copy from QueryPersistenceEngine in Entrance
 */
object JobHistoryService extends Logging {

  private val MAX_DESC_LEN: Int = EngineConnExecutorConfiguration.ERROR_CODE_DESC_LEN

  private val sender: Sender =
    Sender.getSender(EngineConnExecutorConfiguration.JOBHISTORY_SPRING_APPLICATION_NAME.getValue)

  def updateStatus(jobId: Long, status: String): Unit = {
    logger.info("Change job [{}] status to {}", jobId, status)
    if (jobId != null) {}
    val jobRequest: JobRequest = new JobRequest
    jobRequest.setId(jobId)
    jobRequest.setStatus(status)
    update(jobRequest)
  }

  def update(jobRequest: JobRequest): Unit = {
    if (null == jobRequest) { return }
    if (null != jobRequest.getErrorDesc) {
      // length of errorDesc must be less then 1000 / 3
      // because length of errorDesc in db is 1000
      if (jobRequest.getErrorDesc.length > MAX_DESC_LEN) {
        jobRequest.setErrorDesc(jobRequest.getErrorDesc.substring(0, MAX_DESC_LEN))
      }
    }
    jobRequest.setUpdatedTime(new Date)
    val jobReqUpdate: JobReqUpdate = new JobReqUpdate(jobRequest)
    sendToJobHistoryAndRetry(
      jobReqUpdate,
      "job:" + jobRequest.getReqId + "status:" + jobRequest.getStatus
    )
  }

  private def sendToJobHistoryAndRetry(jobReq: RequestProtocol, msg: String): JobRespProtocol = {
    var jobRespProtocol: JobRespProtocol = null
    var retryTimes: Int = 0
    var retry: Boolean = true
    val retryMaxTime = EngineConnExecutorConfiguration.JOBINFO_UPDATE_RETRY_MAX_TIME.getHotValue
    val retryInterval = EngineConnExecutorConfiguration.JOBINFO_UPDATE_RETRY_INTERVAL.getHotValue
    while (retry && retryTimes < retryMaxTime) {
      try {
        retryTimes += 1
        jobRespProtocol = sender.ask(jobReq).asInstanceOf[JobRespProtocol]
        if (jobRespProtocol.getStatus == 2) {
          logger.error(
            s"Request jobHistory failed, joReq msg $msg, retry times: $retryTimes, reason ${jobRespProtocol.getMsg}"
          )
        } else {
          retry = false
        }
      } catch {
        case e: Exception =>
          logger.error(s"Request jobHistory failed, joReq msg $msg, retry times: $retryTimes", e)
      }
      if (retry) {
        Utils.sleepQuietly(retryInterval)
      }
    }
    if (jobRespProtocol != null) {
      val status: Int = jobRespProtocol.getStatus
      val message: String = jobRespProtocol.getMsg
      if (status != 0) {
        logger.warn("Request jobHistory failed,because:{} (请求jobHistory失败)", message)
      }
    } else {
      logger.warn(
        "Request jobHistory failed, because:jobRespProtocol is null (请求jobHistory失败,因为jobRespProtocol为null)"
      )
    }
    jobRespProtocol
  }

}

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/JobHistoryService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.manager.am.service;

import org.apache.linkis.governance.common.entity.job.JobRequest;
import org.apache.linkis.governance.common.protocol.job.JobReqUpdate;
import org.apache.linkis.governance.common.protocol.job.JobRespProtocol;
import org.apache.linkis.manager.am.conf.AMConfiguration;
import org.apache.linkis.protocol.message.RequestProtocol;
import org.apache.linkis.rpc.Sender;

import java.util.Date;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** From Entrance */
public class JobHistoryService {

  public static final JobHistoryService INSTANCE = new JobHistoryService();
  private static final int MAX_DESC_LEN = AMConfiguration.ERROR_CODE_DESC_LEN;
  private Sender sender;

  private static final Logger logger = LoggerFactory.getLogger(JobHistoryService.class);

  private JobHistoryService() {
    sender = Sender.getSender(AMConfiguration.JOBHISTORY_SPRING_APPLICATION_NAME.getValue());
  }

  public void updateStatus(Long jobId, String status, String msg) {
    logger.info("Change job [{}] status to {}", jobId, status);
    JobRequest jobRequest = new JobRequest();
    jobRequest.setId(jobId);
    jobRequest.setStatus(status);
    jobRequest.setErrorDesc(msg);
    update(jobRequest);
  }

  public void update(JobRequest jobRequest) {
    if (null == jobRequest) {
      return;
    }
    if (null != jobRequest.getErrorDesc()) {
      // length of errorDesc must be less then 1000 / 3, because length of errorDesc in db is
      // 1000
      if (jobRequest.getErrorDesc().length() > MAX_DESC_LEN) {
        jobRequest.setErrorDesc(jobRequest.getErrorDesc().substring(0, MAX_DESC_LEN));
      }
    }
    jobRequest.setUpdatedTime(new Date());
    JobReqUpdate jobReqUpdate = new JobReqUpdate(jobRequest);
    sendToJobHistoryAndRetry(
        jobReqUpdate, "job:" + jobRequest.getReqId() + "status:" + jobRequest.getStatus());
  }

  private JobRespProtocol sendToJobHistoryAndRetry(RequestProtocol jobReq, String msg) {
    JobRespProtocol jobRespProtocol = null;
    int retryTimes = 0;
    boolean retry = true;
    while (retry && retryTimes < AMConfiguration.JOBINFO_UPDATE_RETRY_MAX_TIME.getHotValue()) {
      try {
        retryTimes++;
        jobRespProtocol = (JobRespProtocol) sender.ask(jobReq);
        if (jobRespProtocol.getStatus() == 2) {
          logger.warn(
              "Request jobHistory failed, joReq msg{}, retry times: {}, reason {}",
              msg,
              retryTimes,
              jobRespProtocol.getMsg());
        } else {
          retry = false;
        }
      } catch (Exception e) {
        logger.warn(
            "Request jobHistory failed, joReq msg{}, retry times: {}, reason {}",
            msg,
            retryTimes,
            e);
      }
      if (retry) {
        try {
          Thread.sleep(AMConfiguration.JOBINFO_UPDATE_RETRY_INTERVAL.getHotValue());
        } catch (Exception ex) {
          logger.warn(ex.getMessage());
        }
      }
    }
    if (jobRespProtocol != null) {
      int status = jobRespProtocol.getStatus();
      String message = jobRespProtocol.getMsg();
      if (status != 0) {
        logger.warn("Request jobHistory failed,because:{} (请求jobHistory失败)", message);
      }
    } else {
      logger.warn(
          "Request jobHistory failed, because:jobRespProtocol is null (请求jobHistory失败,因为jobRespProtocol为null)");
    }
    return jobRespProtocol;
  }
}
